crypto/cipher_extra/e_tls.c

```diff
@@ -45,19 +45,12 @@ typedef struct {
 OPENSSL_STATIC_ASSERT(EVP_MAX_MD_SIZE < 256,
                       mac_key_len_does_not_fit_in_uint8_t)
 
-OPENSSL_STATIC_ASSERT(sizeof(((EVP_AEAD_CTX *)NULL)->state) >=
-                          sizeof(AEAD_TLS_CTX),
-                      AEAD_state_is_too_small)
-#if defined(__GNUC__) || defined(__clang__)
-OPENSSL_STATIC_ASSERT(alignof(union evp_aead_ctx_st_state) >=
-                          alignof(AEAD_TLS_CTX),
-                      AEAD_state_has_insufficient_alignment)
-#endif
-
 static void aead_tls_cleanup(EVP_AEAD_CTX *ctx) {
-  AEAD_TLS_CTX *tls_ctx = (AEAD_TLS_CTX *)&ctx->state;
+  AEAD_TLS_CTX *tls_ctx = (AEAD_TLS_CTX *)ctx->state.ptr;
   EVP_CIPHER_CTX_cleanup(&tls_ctx->cipher_ctx);
   HMAC_CTX_cleanup(&tls_ctx->hmac_ctx);
```

Comment:
Can't you do with making the `hmac_ctx` field a pointer? And only allocate it when you need it, according to the requested cipher? e.g. `aes-gcm` won't need it.

Reply:
The casting was getting annoying so I just added a `(void*)` pointer to the union. Since that's smaller than the existing union it doesn't actually change anything, just makes the casting much easier.

Reply:
I was thinking about avoiding dynamic memory allocation https://github.com/awslabs/aws-lc/pull/368/files#diff-544b0ce33ba4678dc2dcd3e8667b8473e4e5cf2e67291ea089e5ab3964d3cdb5R76

unless necessary depending on the cipher. Might not be technically possible, of course. Just thoughts.

Reply:
Actually, I think that this code-path is only encountered when HMAC is needed, so we may be fine.

Reply:
Ahh that's correct. I confused it with the APIs the layer above this one!